File: 2364_count_number_of_bad_pairs.cpp
#include <unordered_map>
#include <vector>

class Solution {
public:
    long long countBadPairs(std::vector<int>& nums) {
        long long n = nums.size();
        long long totalPairs = n * (n - 1) / 2;
        long long goodPairs = 0;
        std::unordered_map<int, int> freqMap;

        for (int i = 0; i < n; i++) {
            int key = nums[i] - i;
            if (freqMap.count(key)) {
                goodPairs += freqMap[key];
            }
            freqMap[key]++;
        }
        return totalPairs - goodPairs;
    }
};

File: 2364_count_number_of_bad_pairs.java
import java.util.*;

class Solution {
    public long countBadPairs(int[] nums) {
        long n = nums.length;
        long totalPairs = n * (n - 1) / 2;
        long goodPairs = 0;
        Map<Integer, Integer> countMap = new HashMap<>();

        for (int i = 0; i < n; i++) {
            int key = nums[i] - i;
            goodPairs += countMap.getOrDefault(key, 0);
            countMap.put(key, countMap.getOrDefault(key, 0) + 1);
        }
        return totalPairs - goodPairs;
    }
}
